transports/quic/src/lib.rs

```diff
@@ -0,0 +1,82 @@
+// Permission is hereby granted, free of charge, to any person obtaining a
+// copy of this software and associated documentation files (the "Software"),
+// to deal in the Software without restriction, including without limitation
+// the rights to use, copy, modify, merge, publish, distribute, sublicense,
+// and/or sell copies of the Software, and to permit persons to whom the
+// Software is furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS
+// OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+// FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
+// DEALINGS IN THE SOFTWARE.
+
+//! Implementation of the libp2p `Transport` and `StreamMuxer` traits for QUIC.
+//!
+//! # Usage
+//!
+//! Example:
+//!
+//! ```
+//! # #[cfg(not(feature = "async-std"))]
+//! # fn main() {}
+//! #
+//! # #[cfg(feature = "async-std")]
+//! # fn main() -> std::io::Result<()> {
+//! #
+//! use libp2p_quic::{AsyncStdTransport, Config};
+//! use libp2p_core::{Multiaddr, Transport};
+//!
+//! let keypair = libp2p_core::identity::Keypair::generate_ed25519();
+//! let quic_config = Config::new(&keypair).expect("could not make config");
+//!
+//! let mut quic_transport = AsyncStdTransport::new(quic_config);
+//!
+//! let addr = "/ip4/127.0.0.1/udp/12345/quic".parse().expect("bad address?");
+//! quic_transport.listen_on(addr).expect("listen error.");
+//! #
+//! # Ok(())
+//! # }
+//! ```
+//!
+//! The [`Endpoint`] struct implements the [`Transport`] trait of the [`libp2p_core`] library. See the
+//! documentation of [`libp2p_core`] and of libp2p in general to learn how to use the [`Transport`] trait.
+//!
+//! Note that QUIC provides transport, security, and multiplexing in a single protocol.  Therefore,
+//! QUIC connections do not need to be upgraded. You will get a compile-time error if you try.
+//! Instead, you must pass all needed configuration into the constructor.
+//!
+//! # Design Notes
+//!
+//! The entry point is the `Endpoint` struct.  It represents a single QUIC endpoint.  You
+//! should generally have one of these per process.
+//!
+//! `Endpoint` manages a background task that processes all incoming packets.  Each
+//! `QuicConnection` also manages a background task, which handles socket output and timer polling.
+
+#![deny(unsafe_code)]
+
+mod connection;
+mod endpoint;
+mod muxer;
+mod tls;
+mod transport;
+mod upgrade;
+
+pub use connection::ConnectionError;
+pub use endpoint::Config;
+pub use muxer::QuicMuxer;
+pub use quinn_proto::ConnectError as DialError;
```

Comment:
`quinn_proto::ConnectError` is currently wrapped as a variant of `ConnectionError`. We could remove it and instead inline each variant of `quinn_proto::ConnectError` into `ConnectionError`, but I don't think its worth the effort just to avoid the re-export.

Reply:
I've had a quick (pun intended) look and this seems to be almost the only case where we have `quinn_proto` types in our public API.

Not re-exporting it would free us from having to make a major bump every time `quinn_proto` updates.

Any easy way of handling this with errors is be wrapping them in a newtype and slapping `#[error(transparent)]` on it.

Reply:
Added single error type for both connection and transport, and newtypes for `quinn_proto` errors. See c0b8a55c100346c49b4428dca8642304f6238c5f. Wdyt @thomaseizinger?